Introduce pagination in files-filter report

DeepDiver1975 opened the pull request:
## Description
Allow pagination when listing files

### ToDos
- [x] integration tests
- [ ] add cache for file list
- [ ] integrate on client side

## Motivation and Context

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->

<!--- Include details of your testing environment, and the tests you ran to -->

<!--- see how your change affects other areas of the code, etc. -->
## Screenshots (if appropriate):
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

DeepDiver1975 committed:
Introduce pagination in files-filter report

PVince81 commented:
for https://github.com/owncloud/core/issues/13915

PVince81 commented:
>  add cache for file list

memcache ? 😦

PVince81 commented on apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:
```diff
@@ -158,51 +161,44 @@ public function getSupportedReportSet($uri) {
-					$requestedProps[] = $propVal['name'];
-				}
+		$requestedProps = $report->properties;
+		$filterRules = $report->filters;
```
really ? did the format change ?

PVince81 commented on apps/dav/lib/Files/Xml/FilterRequest.php:
```diff
@@ -0,0 +1,104 @@
+use Sabre\Xml\Reader;
+use Sabre\Xml\XmlDeserializable;
+
+class FilterRequest implements XmlDeserializable {
```
oohhhh, clever

PVince81 commented:
Code looks good so far

PVince81 commented on apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:
```diff
@@ -158,51 +161,44 @@ public function getSupportedReportSet($uri) {
-			throw new PreconditionFailed('Cannot filter by non-existing tag', 0, $e);
+		if (!is_null($report->limit)) {
+			$length = $report->limit['size'];
+			$offset = $report->limit['page'] * $length;
```
The first page should be 1, not 0.

How about using limit and offset instead ? Might be easier to grasp...

PVince81 commented:
Test report report.xml:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<oc:filter-files xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns" >
        <a:prop>
                <a:getetag/>
        </a:prop>
        <oc:limit page="0" size="10" />
</oc:filter-files>
```
```bash
% curl -u admin:admin -X REPORT -H "Content-Type: text/xml" --data-binary "@report.xml" 'http://localhost/owncloud/remote.php/dav/files/admin/'
```

PVince81 commented:
Some concerns: the regular filter report returns a **flat** list of files from all folders. This flat list can be paginated in this report, this is fine.

However we will also need a paginated call for folder-specific contents, without recursing. Should we make a separate report time for this maybe ? Or add a property `<oc:recursive />` to make it a flat list, and by default only return for the current folder ?

PVince81 commented:
@DeepDiver1975 I've rebased this.

Also I discovered today that Sabre has a shorter way of compiling results from multiple nodes using `getPropertiesForMultiplePaths`. Since we're touching this report anyway, I added it here f4aa5d29df4c382a7ac89a95dbbf3b96b682f4e3

There will still be a performance issue here because we need to resolve every node individually.
Ideally for pagination we might already know when we're fetching from a single folder, so this fetching should be optimizable. Might be doable using the `IMultiget` interface which `getPropertiesForMultiplePaths` uses internally: https://github.com/fruux/sabre-dav/blob/master/lib/DAV/IMultiGet.php#L34. When implementing this, we could sort the results by folder and optimize fetching props of specific grouped folders.

Side note: if we ever implement Webdav sync it also uses `IMultiGet` 😉

PVince81 commented:
- [ ] sort out "recursive" vs "flat files" question: https://github.com/owncloud/core/pull/26507#issuecomment-267968812
- [ ] properties resolving perf: https://github.com/owncloud/core/pull/26507#issuecomment-268033023
- [ ] fix page number to start with 1: https://github.com/owncloud/core/pull/26507#discussion_r93034201
- [ ] use `DAV:limit` for the limit ? (see https://tools.ietf.org/search/rfc5323#section-5.17)

PVince81 commented:
Arghh, no, not good. `getPropertiesForMultiplePaths` would redo a `getChild()` on the parent nodes because we haven't created any Sabre node yet for all the results.

I suspect that the old implementation also suffered from this. Since we have the files API node already, we should already be able to wrap it (it's a `FileInfo` after all) into a Sabre node and treat it directly as a result. No need for DB refetching...

I'll revert my commit because it will degrade the performance. From what I see we already had created the matching Sabre File and Directory in `findNodesByFileIds`.

More about this in a minute after the revert...

PVince81 commented:
So, I've reverted the commit instead of deleting it, in case we still want it. (Squashing will kill it later).

Now looking at 'findNodesByFileIds()`: this has BAD performance because it will resolve every fileid using `Folder::getById()`. Now for arbitrary results like "/test/x1", "/test/sub/x2", "/abc/def" it makes sense as we're not in the same folder. This was valid for the "filter by tags" and "filter by favorite" feature in which results are usually not inside the same folder.

However if we want REPORT to be used for pagination on simple folders, then we need a different approach there. Usually calling `getChildren()` on the parent is more efficient as it does a single DB query for `getDirectoryContents()`.

So to solve the above question we really need to discuss "recursive / flattened / arbitrary results" vs "only show results from the current folder".

One idea would be to provide one generic REPORT type that only does a PROPFIND on the files endpoints (or even any part of the tree!) and paginates whatever the PROPFIND / `getChildren()` returns. All results are children of the current node.

Then we provide another REPORT, which is the one from this PR, to really do search-like queries. In such queries the results are not limited to the current folder, so we **have** to use `getById()` for every result.

Or a hybrid variant is to have a single REPORT type but with a `oc:recursive` property. When unset, we'd resort to a simple `getChildren()` + post-filter on the results. When set, we'd do the other approach that gets the file ids of the results first, then resolve then. Note that both implementations are completely different.

@DeepDiver1975 thoughts ?

PVince81 commented:
I wrote down some ideas about implementing a generic pagination report that would work for every PROPFIND-enabled endpoint: https://github.com/owncloud/core/issues/26840

PVince81 commented:
Regarding recursive vs children, I think we could even call them differently:
- recursive: is more a SEARCH operation because you're searching for all nodes that match a specific criteria in all subcollections. So the approach is more like asking an index (if we had one) to give results in many folders and then compile these results in one result set.
- children: is more of a filter where you basically just get the contents of the current collection (children) and then apply a filter on their properties. All data is already there within the current collection, so no need for an index, just need to remove the nodes from the result that do not match the filter criteria.

PVince81 commented:
So for "recursive" vs "flat" the answer from @DeepDiver1975 was using the "Depth" header.
Which means that to search on a single collection one specifies "Depth: 1" and in all subcollections "Depth: infinite".

Commit:
Use a more API-style way to generate REPORT PROPFIND

This will also properly return 404 properties

PVince81 commented:
Rebased.

Also I found a way to simplify the response code to use a more Sabr-y way to do it: https://github.com/owncloud/core/pull/26507/commits/c355695cba542acffcb05f25e19532a202557e98

All discovered while working on the customgroups REPORT: https://github.com/owncloud/customgroups/pull/27.

Note: in the customgroups pagination I used a different format for the search pattern + pagination:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<oc:search-query xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns">
        <a:prop>
                <oc:role></oc:role>
                <oc:display-name></oc:display-name>
        </a:prop>
        <oc:search>
                <oc:pattern>us</oc:pattern>
                <oc:limit>5</oc:limit>
                <oc:offset>0</oc:offset>
        </oc:search>
</oc:search-query>
```

PVince81 commented:
How about we finish this without the cache ? It will be crude but at least it will make it possible for clients to at least request the first page with limit.

Commit:
Adjust REPORT search query, perf and unit tests

- disable REPORT without filter

Because we need to use Node::search($pattern) to find all matching nodes
in all the subfolder recursively, but the result nodes contain
incomplete information like owner.

- remove unneeded trailing slash when buildling response href

This got obsoleted when switching to generateMultiStatus()

- add integration pagination test for favorites REPORT

- throw exception for unknown node types in FilesReportPlugin

PVince81 commented:
Rebased and pushed some changes:

- pagination (offset+limit) now works also for tags/favorite
- disabled case where no filter is specified because `getChildren()` was not enough and would only search one level deep. Actually we need to use `Node::search()` but when I tried I noticed that it was missing attributes in its FileInfo properties (like owner) which makes it currently unusable. Something to dig into in the future (<= @DeepDiver1975)
- added unit tests

Next up: integration tests for pagination when filtering.

PVince81 commented:
```
21:23:41     /var/lib/jenkins/workspace/owncloud-core_core_PR-26507-XBSZ3PUQKGNP3NP62UQHR4G7JKGONNKXFRY3NXEAITQYFZVK5JDQ/tests/integration/features/favorites.feature:81
21:23:41     /var/lib/jenkins/workspace/owncloud-core_core_PR-26507-XBSZ3PUQKGNP3NP62UQHR4G7JKGONNKXFRY3NXEAITQYFZVK5JDQ/tests/integration/features/favorites.feature:93
```

PVince81 commented:
Fixed integration tests and also added a new integration test for pagination.

@DeepDiver1975 @jvillafanez please review

jvillafanez commented on apps/dav/lib/Files/Xml/FilterRequest.php:
```diff
@@ -0,0 +1,110 @@
+			}
+		}
+
+		$obj = new self();
```
Any reason to use `self` instead of explicitly calling the current class?

PVince81 replied:
Written by @DeepDiver1975 and likely copied from other Sabre deserializer classes where `self` was used. More convenient than rewriting the current class name I guess.

jvillafanez replied:
If it was copied from Sabre, i guess it's fine. Don't like it, but ok.

jvillafanez commented on tests/integration/features/bootstrap/WebDav.php:
```diff
@@ -394,18 +394,33 @@ public function listFolder($user, $path, $folderDepth, $properties = null){
-									' . $filterRules . '
-								 </oc:filter-rules>
-							 </oc:filter-files>';
+					<oc:filter-files xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns" >
```
We have to find a better way to generate this.

PVince81 replied:
Agreed. I'll let @owncloud/qa go wild to improve this separately.

jvillafanez commented on apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:
```diff
@@ -378,17 +385,25 @@ public function findNodesByFileIds($rootNode, $fileIds) {
+			return new File($this->fileView, $filesNode);
+		} else if ($filesNode instanceof \OCP\Files\Folder) {
+			return new Directory($this->fileView, $filesNode);
+		}
```
We should log something here unless this `null` is properly detected and reported. I wouldn't like to miss this case if something goes horribly wrong.

PVince81 replied:
I'll throw a "WTF this can't be" style exception

jvillafanez replied:
Any chance that this exception is cached in the middle and doesn't show up? just making sure this will be visible.

PVince81 replied:
It's wrapped in a SabreException later upstream:
```
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Exception</s:exception>
  <s:message>Unrecognized Files API node returned, aborting</s:message>
</d:error>
```

jvillafanez commented:
I haven't checked deeply the FilesReportPlugin.php file, in case someone wants to recheck that file

PVince81 commented on apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:
```diff
@@ -159,58 +162,71 @@ public function getSupportedReportSet($uri) {
-		}
+			// TODO: if depth one, could use a different search algo based on getChildren() ...
+			// TODO: Depth header filtering?
+			$results = $folder->search($report->search['pattern']);
```
boohoo, ok I'll move it to another WIP PR

PVince81 commented on apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:
```diff
@@ -159,58 +162,71 @@ public function getSupportedReportSet($uri) {
+		$requestedProps = $report->properties;
+		$filterRules = $report->filters;
+
+		if (empty($filterRules['systemtag']) && is_null($filterRules['favorite'])) {
```
Not sure... "systemtag" here is an array so empty fits but "favorite" is only a single string.

jvillafanez replied:
if "systemtag" will be always an array no matter what, then it's fine. Something similar for "favorite" (wondering if `''` or '0' would behave as we expect).
Anyway a comment would be useful to clarify any doubt, mainly to say that this is on purpose.

PVince81 commented:
> I haven't checked deeply the FilesReportPlugin.php file, in case someone wants to recheck that file

There aren't many changes there so I think it's ok.

Basically the response stays the same except that the array we pass there is sliced.
Most of the added logic is to slice at the right moment.

@jvillafanez adjusted, see last two commits

PVince81 commented:
Final review ? @jvillafanez @DeepDiver1975 
Thanks.

jvillafanez commented:
https://github.com/owncloud/core/pull/26507#discussion_r107402991 and https://github.com/owncloud/core/pull/26507#discussion_r107404193 . Other than that 👍

PVince81 commented:
Added comment, squashed my commits into a smaller set.

Since there are no actual code changes -> instamerge

PVince81 commented:
Raised doc ticket with some instructions: https://github.com/owncloud/documentation/issues/2950
